keras/layers/layer.py

```diff
@@ -1060,23 +1060,24 @@ def call(self, x):
         else:
             self._losses.extend(losses)
 
-    def _get_own_losses(self):
+    def _get_added_losses(self):
+        """List of scalar losses from `add_loss` including from sublayers."""
         if backend.in_stateless_scope():
             losses = []
             scope = backend.get_stateless_scope()
             for loss in scope.losses:
                 if id(loss) in self._loss_ids:
                     losses.append(loss)
-            return losses
         else:
-            return self._losses[:]
+            losses = list(self._losses[:])
+        for layer in self._layers:
+            losses.extend(layer._get_added_losses())
+        return losses
 
     @property
     def losses(self):
-        """List of scalar losses added via `add_loss()` during layer call."""
-        losses = self._get_own_losses()
-        for layer in self._layers:
```

Comment:
To reduce redundancy and improve recursive traversal efficiency you could just replace `self._layers` with `self._flatten_layers(include_self=False)` and leave `_get_own_losses` unchanged.

I wonder whether there was a reason we didn't make it recursive or whether it was just an oversight. Seems like it should be recursive.